### PR TITLE
Norm-fiber op for batchnorm2d

### PR DESCRIPTION
**Pull request by gogolgrind**
work in progress

### CODE REVIEW

File: include/nntile/tile/norm_fiber.hh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <nntile/tile/tile.hh>
+
+namespace nntile::tile
+{
+
+// Tile-wise sum_fiber
+template<typename T>
+void norm_fiber_async(Scalar alpha, const Tile<T> &src, Scalar beta, const Tile<T> &dst,
+        Index axis, Index batch_ndim);
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:25]: Add redux parameter. You may ignore it, just add it to API of this unction

File: include/nntile/tile/norm_fiber.hh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+// Tile-wise sum_fiber
+template<typename T>
+void norm_fiber(Scalar alpha, const Tile<T> &src, Scalar beta, const Tile<T> &dst,
+        Index axis, Index batch_ndim);
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:30]: The same with redux flag.

File: src/kernel/norm_fiber/cpu.cc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+template
+void cpu<fp32_t>(Index m, Index n, Index k, Index batch, Scalar alpha,
+        const fp32_t *src, Scalar beta, fp32_t *dst)
+    noexcept;
+
+template
+void cpu<fp64_t>(Index m, Index n, Index k, Index batch, Scalar alpha,
+        const fp64_t *src, Scalar beta, fp64_t *dst)
+    noexcept;
+
```
Comment (Muxas) [inline src/kernel/norm_fiber/cpu.cc:141]: cpu<fp32_fast_tf32_t> is missing

File: src/kernel/norm_fiber/cuda.cu [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+template
+void cuda<fp32_t>(cudaStream_t stream, Index m, Index n, Index k, Index batch,
+        Scalar alpha, const fp32_t *src, Scalar beta, fp32_t *dst)
+    noexcept;
+
+template
+void cuda<fp64_t>(cudaStream_t stream, Index m, Index n, Index k, Index batch,
+        Scalar alpha, const fp64_t *src, Scalar beta, fp64_t *dst)
+    noexcept;
+
```
Comment (Muxas) [inline src/kernel/norm_fiber/cuda.cu:99]: cuda<fp32_fast_tf32_t> is missing

File: src/starpu/norm_fiber.cc [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+#ifdef NNTILE_USE_CUDA
+            {cuda<bf16_t>}
+#else // NNTILE_USE_CUDA
+            {}
+#endif // NNTILE_USE_CUDA
+            );
+
+    codelet_fp32_fast_tf32.init("nntile_norm_fiber_fp32_fast_tf32",
+            footprint,
+            {cpu<fp32_t>},
```
Comment (Muxas) [inline src/starpu/norm_fiber.cc:105]: <fp32_fast_tf32_t>

File: src/starpu/norm_fiber.cc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+#ifdef NNTILE_USE_CUDA
+            {cuda<fp32_t>}
```
Comment (Muxas) [inline src/starpu/norm_fiber.cc:107]: <fp32_fast_tf32_t>

File: src/tile/norm_fiber.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+#include "nntile/tile/norm_fiber.hh"
+#include "nntile/starpu/norm_fiber.hh"
+
+namespace nntile::tile
+{
+
+template<typename T>
+void norm_fiber_async(Scalar alpha, const Tile<T> &src, Scalar beta,
+        const Tile<T> &dst, Index axis, Index batch_ndim)
```
Comment (Muxas) [inline src/tile/norm_fiber.cc:23]: add redux flag

File: include/nntile/tile/norm_fiber.hh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <nntile/tile/tile.hh>
+
+namespace nntile::tile
+{
+
+// Tile-wise norm_fiber
+template<typename T>
+void norm_fiber_async(Scalar alpha, const Tile<T> &src, Scalar beta, const Tile<T> &dst,
+        Index axis, Index batch_ndim, int redux);
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:25]: `redux=0` for a default value, as in src/tensor/norm_fiber.hh

File: include/nntile/tile/norm_fiber.hh [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+// Tile-wise norm_fiber
+template<typename T>
+void norm_fiber(Scalar alpha, const Tile<T> &src, Scalar beta, const Tile<T> &dst,
+        Index axis, Index batch_ndim, int redux);
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:30]: `redux=0`

File: src/kernel/norm_fiber/cuda.cu [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    Index i2 = i2_batched % k;
+    Index b = i2_batched / k;
+    Index i0_start = threadIdx.y, i0_step = blockDim.y;
+    Index i1_start = threadIdx.z, i1_step = blockDim.z;
+    using Y = typename T::repr_t;
+    constexpr Y zero{0.0};
+    // Init sum
+    Y sum = zero;
+    */
+    // this is placholder - TODO
```
Comment (Muxas) [inline src/kernel/norm_fiber/cuda.cu:56]: cuda code does nothing, is it intended?

File: src/tensor/norm_fiber.cc [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    if(dst.shape[0] != src.shape[axis])
+    {
+        throw std::runtime_error("dst.shape[0] != src.shape[axis]");
+    }
+    if(dst.basetile_shape[0] != src.basetile_shape[axis])
+    {
+        throw std::runtime_error("dst.basetile_shape[0] != "
+                "src.basetile_shape[axis]");
+    }
+    for(Index i = 0; i < batch_ndim; ++i)
```
Comment (Muxas) [inline src/tensor/norm_fiber.cc:55]: This cycle is incorrect. If `batch_ndim=0`, then this check does nothing. There must be 2 separate check cycles, take a look at `sum_fiber` function.

File: src/tile/norm_fiber.cc [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    if(axis >= ndim-batch_ndim)
+    {
+        throw std::runtime_error("axis >= ndim-batch_ndim");
+    }
+    // Check shapes
+    if(dst.shape[0] != src.shape[axis])
+    {
+        throw std::runtime_error("dst.shape[0] != src.shape[axis]");
+    }
+    for(Index i = 0; i < batch_ndim; ++i)
```
Comment (Muxas) [inline src/tile/norm_fiber.cc:50]: Wrong check of shapes, check `sum_fiber`

**Review by Muxas — CHANGES_REQUESTED** [review]

File: src/kernel/norm_fiber/cuda.cu (hunk 10 above)
Comment (gogolgrind) [reply src/kernel/norm_fiber/cuda.cu:56]: Yes - as we disused cuda will be added later

File: include/nntile/tile/norm_fiber.hh [hunk 13]
```diff
@@ -0,0 +1,32 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+ *                              (Skoltech), Russia. All rights reserved.
+ *                 2023-present Artificial Intelligence Research Institute
+ *                              (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on StarPU runtime system.
+ *
+ * @file include/nntile/tile/sum_fiber.hh
+ * Euclidean norms over slices into a fiber of  of a Tile<T>
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:10]: `of  of`

File: include/nntile/tile/norm_fiber.hh [hunk 14]
```diff
@@ -0,0 +1,32 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+ *                              (Skoltech), Russia. All rights reserved.
+ *                 2023-present Artificial Intelligence Research Institute
+ *                              (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on StarPU runtime system.
+ *
+ * @file include/nntile/tile/sum_fiber.hh
```
Comment (Muxas) [inline include/nntile/tile/norm_fiber.hh:9]: `sum_fiber` -> `norm_fiber`

File: src/kernel/norm_fiber/cpu.cc [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+                // Get sum of a corresponding slice
+                const T *src_slice = src + ((i1+b*n)*k+i2)*m;
+                // Cycle over the first axis of input buffer
+                for(Index i0 = 0; i0 < m; ++i0)
+                {
+                    // Read value from source
+                    Y val = std::fabs(Y{src_slice[i0]});
+                    // Update norm only if new value is non-zero
+                    if(val > 0)
+                    {
```
Comment (Muxas) [inline src/kernel/norm_fiber/cpu.cc:72]: Add a comment that we use Kahan summation rule to get scaled sum of squares

File: src/kernel/norm_fiber/cpu.cc [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+                            c = (t-norm_ssq) - y;
+                            norm_ssq = t;
+                            norm_max = val;
+                        }
+                    }
+                }
+            }
+            // Get the scaled norm
+            norm_max *= alpha;
+            //T norm = norm_max * std::sqrt(norm_ssq);
```
Comment (Muxas) [inline src/kernel/norm_fiber/cpu.cc:97]: rm this line

File: src/kernel/norm_fiber/cpu.cc [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            // Update output value
+            if(beta == zero)
+            {
+                //result = norm;
```
Comment (Muxas) [inline src/kernel/norm_fiber/cpu.cc:101]: rm this line

File: src/kernel/norm_fiber/cpu.cc [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+                result = static_cast<T>(norm_max * std::sqrt(norm_ssq));
+            }
+            else if(norm_max > 0)
+            {
+                //result = std::hypot(beta*result, norm);
```
Comment (Muxas) [inline src/kernel/norm_fiber/cpu.cc:106]: rm this line

File: wrappers/python/nntile/layer/batch_norm.py
Comment (Muxas) [inline wrappers/python/nntile/layer/batch_norm.py:1]: Merge main from the main repository into this PR to get the latest version of this file in main

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 19]
```diff
@@ -171,4 +171,4 @@ def test_backward(self, starpu_simple, numpy_rng, torch_rng,
             nntile.tensor.to_numpy(nntile_layer.x.grad),
             input_torch.grad.numpy(),
             atol=params.atol,
-        )
+        )
```
Comment (Muxas) [inline wrappers/python/tests/layer/test_batch_norm.py:174]: no end of line at the end of the file

File: src/kernel/norm_fiber/cuda.cu [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+ * */
+{
+    Index i2_batched = threadIdx.x + blockIdx.x*blockDim.x;
+    Index i2 = i2_batched % k;
+    Index b = i2_batched / k;
+    Index i0_start = threadIdx.y, i0_step = blockDim.y;
+    Index i1_start = threadIdx.z, i1_step = blockDim.z;
+    using Y = typename T::repr_t;
+    constexpr Y zero{0.0};
+    // WORK IN PROGRESS!!!
```
Comment (Muxas) [inline src/kernel/norm_fiber/cuda.cu:53]: Is it still in progress?

File: src/kernel/norm_fiber/cuda.cu [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            // Get sum of a corresponding slice
+            const T *src_slice = src + ((i1+b*n)*k+i2)*m;
+            // Cycle over the first axis of input buffer
+            for(Index i0 = i0_start; i0 < m; i0 += i0_step)
+            {
+                // Read value from source and Update sum
+                sum = ::hypot(sum, Y{src_slice[i2*m]});
+            }
+        }
+        // WORK IN PROGRESS!!!
```
Comment (Muxas) [inline src/kernel/norm_fiber/cuda.cu:70]: Remove redundant comments

File: wrappers/python/nntile/layer/batch_norm.py [hunk 22]
```diff
@@ -44,22 +46,18 @@ def __init__(
         self.y = y
 
         self.x_unbiased_copy = zeros(self.y.value.shape, dtype=self.dtype)
```
Comment (Muxas) [inline wrappers/python/nntile/layer/batch_norm.py:48]: `zeros`->`empty` everywhere

**Review by Muxas — CHANGES_REQUESTED** [review]

**Review by Muxas — APPROVED** [review]
Just fix linting and check if all tests pass
